api/apps/api/src/modules/projects/project-cost-surface.controller.ts

```diff
@@ -0,0 +1,132 @@
+import {
+  Controller,
+  forwardRef,
+  Get,
+  Header,
+  Inject,
+  Param,
+  ParseUUIDPipe,
+  Post,
+  Req,
+  Res,
+  UploadedFile,
+  UseGuards,
+} from '@nestjs/common';
+import { ApiOkResponse, ApiOperation, ApiTags } from '@nestjs/swagger';
+import { projectResource } from '@marxan-api/modules/projects/project.api.entity';
+import { apiGlobalPrefixes } from '@marxan-api/api.config';
+import { ApiConsumesShapefile } from '@marxan-api/decorators/shapefile.decorator';
+import {
+  GeometryFileInterceptor,
+  GeometryKind,
+} from '@marxan-api/decorators/file-interceptors.decorator';
+import { asyncJobTag } from '@marxan-api/dto/async-job-tag';
+import { RequestWithAuthenticatedUser } from '@marxan-api/app.controller';
+import {
+  AsyncJobDto,
+  JsonApiAsyncJobMeta,
+} from '@marxan-api/dto/async-job.dto';
+import { ensureShapefileHasRequiredFiles } from '@marxan-api/utils/file-uploads.utils';
+import { isLeft } from 'fp-ts/Either';
+import { mapAclDomainToHttpError } from '@marxan-api/utils/acl.utils';
+import { scenarioResource } from '@marxan-api/modules/scenarios/scenario.api.entity';
+import { ScenariosService } from '@marxan-api/modules/scenarios/scenarios.service';
+import { CostRangeDto } from '@marxan-api/modules/scenarios/dto/cost-range.dto';
+import { plainToClass } from 'class-transformer';
+import { Response } from 'express';
+import { marxanRunFiles } from '@marxan-api/modules/scenarios/scenarios.controller';
+import { JwtAuthGuard } from '@marxan-api/guards/jwt-auth.guard';
+import { ImplementsAcl } from '@marxan-api/decorators/acl.decorator';
+
+@ApiTags(projectResource.className)
+@Controller(`${apiGlobalPrefixes.v1}/projects`)
+export class ProjectCostSurfaceController {
+  constructor(
+    @Inject(forwardRef(() => ScenariosService))
+    public readonly scenarioService: ScenariosService,
+  ) {}
+
+  @ImplementsAcl()
+  @UseGuards(JwtAuthGuard)
+  @ApiConsumesShapefile({ withGeoJsonResponse: false })
+  @GeometryFileInterceptor(GeometryKind.ComplexWithProperties)
+  @ApiTags(asyncJobTag)
+  @Post(`:scenarioId/cost-surface/shapefile`)
+  async processCostSurfaceShapefile(
+    @Param('scenarioId') scenarioId: string,
+    @Req() req: RequestWithAuthenticatedUser,
+    @UploadedFile() file: Express.Multer.File,
+  ): Promise<JsonApiAsyncJobMeta> {
+    await ensureShapefileHasRequiredFiles(file);
+
+    const result = await this.scenarioService.processCostSurfaceShapefile(
+      scenarioId,
+      req.user.id,
+      file,
+    );
+
+    if (isLeft(result)) {
+      throw mapAclDomainToHttpError(result.left, {
+        scenarioId,
+        userId: req.user.id,
+        resourceType: scenarioResource.name.plural,
+      });
+    }
+    return AsyncJobDto.forScenario().asJsonApiMetadata();
+  }
```

Comment:
for this handler, we'll need to coordinate with frontend and accept a `projectId` instead, and then move the `processCostSurfaceShapefile()` method to a different service (moving it out of `ScenariosService`, basically, as it won't belong there anymore).

If we want to break the fewest things possible, we may want to keep this method and its current code path in `ScenariosController` while we complete the basic workflow for id26.2, so that the current FE UI can still allow users to upload cost surfaces directly to a scenario, and then we (including FE UI) can switch over to the "upload to project, link to scenario" workflow once ready

Reply:
Right, I totally missed that in the scope of this task. I was planning to handle MRXN23-89, 23-89 and 23-90 as they felt kind of isolated enough and a good starting point.

I have decided to rollback and deprecate the endpoint in ScenariosController to not disturb any current functionality at then tmp throw a not implemented exception in the new controller. I hope this will adjust better to the scope until I get more understanding of the whole epic.